rust/kernel/fs.rs

```diff
@@ -0,0 +1,126 @@
+pub mod address_space;
+pub mod address_space_operations;
+pub mod dentry;
+pub mod inode;
+pub mod inode_operations;
+pub mod kiocb;
+pub mod libfs_functions;
+pub mod super_block;
+pub mod super_operations;
+
+use core::ptr;
+
+use crate::{
+    bindings, c_types::*, error::from_kernel_err_ptr, fs::super_block::SuperBlock, print::ExpectK,
+    ret_err_ptr, str::CStr, types::FileSystemFlags, Result,
+};
+
+pub trait BuildVtable<T> {
+    fn build_vtable() -> &'static T;
+}
+#[macro_export]
+macro_rules! declare_c_vtable {
+    ($O:ident, $T:ty, $val:expr $(,)?) => {
+        pub struct $O;
+        impl $crate::fs::BuildVtable<$T> for $O {
+            fn build_vtable() -> &'static $T {
+                unsafe { &($val) }
+            }
+        }
+    };
+}
+
+pub type FileSystemType = bindings::file_system_type;
+
+pub trait FileSystemBase {
+    type MountOptions = c_void;
+
+    const NAME: &'static CStr;
+    const FS_FLAGS: FileSystemFlags;
+    const OWNER: *mut bindings::module = ptr::null_mut();
+
+    fn mount(
+        fs_type: &'_ mut FileSystemType,
+        flags: c_int,
+        device_name: &CStr,
+        data: Option<&mut Self::MountOptions>,
+    ) -> Result<*mut bindings::dentry>;
+
+    fn kill_super(sb: &mut SuperBlock);
+
+    fn fill_super(
+        sb: &mut SuperBlock,
+        data: Option<&mut Self::MountOptions>,
+        silent: c_int,
+    ) -> Result;
+}
+
+pub trait DeclaredFileSystemType: FileSystemBase {
+    fn file_system_type() -> *mut bindings::file_system_type;
+}
+
+#[macro_export]
+macro_rules! declare_fs_type {
+    ($T:ty, $S:ident) => {
+        static mut $S: $crate::bindings::file_system_type = $crate::bindings::file_system_type {
```

Comment:
Where is this static mutated? Is it implicitly synchronized, or does it need a mutex?

Reply:
We need a `*mut bindings::file_system_type` for `mount_nodev` and I think we faced problems when this was just a `static` (are normal statics in read-only memory?). In the C impl, this was just a (non-const) global variable and we figured this would be the closest match. The vtable approach from `FileOperations` doesn't work either because the fields are mutated in C code

Maybe we are missing something though

Reply:
IIUC, This is only used for fs registration. It will not be mutated later from fs side.